compiler/src/iree/compiler/Dialect/Stream/Transforms/SpecializeEncodings.cpp

```diff
@@ -0,0 +1,172 @@
+    // TODO(hanchung): We should use the default device in this case. However,
+    // it is not guaranteed that default device attribute will always be set in
+    // the IR. (Is the statement correct?)
+    auto affinityAttr = affinityOp.getAffinityAttr();
+    if (!affinityAttr) {
+      return;
+    }
+    candidates.push_back(affinityOp);
+  });
+
+  if (candidates.empty()) {
+    return success();
+  }
+
+  IRRewriter rewriter(funcOp.getContext());
+  for (auto affinityOp : candidates) {
+    auto affinityAttr = affinityOp.getAffinityAttr();
+    SetVector<Attribute> layouts;
+    if (failed(resolveLayoutAttr(affinityAttr, moduleOp, layouts))) {
+      return affinityOp.emitError("failed on making layouts");
+    }
+
+    // Returns an updated encoding attribute if an encoding attribute is present
+    // in the type. Otherwise, returns std::nullopt.
+    auto getEncodingWithNewLayouts =
+        [=](Type type) -> std::optional<IREE::Encoding::EncodingAttr> {
+      auto rankedTensorType = dyn_cast<RankedTensorType>(type);
+      if (!rankedTensorType) {
+        return std::nullopt;
+      }
+      auto encodingAttr = IREE::Encoding::getEncodingAttr(rankedTensorType);
+      if (!encodingAttr) {
+        return std::nullopt;
+      }
+      return encodingAttr.cloneWithLayouts(layouts.getArrayRef());
+    };
+
+    // TODO(hanchung): Update other Stream operations.
+    LogicalResult result =
+        TypeSwitch<Operation *, LogicalResult>(affinityOp)
+            .Case<IREE::Stream::TensorSizeOfOp>([&](auto sizeOfOp) {
+              auto encodingType =
+                  dyn_cast<RankedTensorType>(sizeOfOp.getEncoding());
+              if (!encodingType) {
+                return success();
+              }
+              std::optional<IREE::Encoding::EncodingAttr> encodingAttr =
+                  getEncodingWithNewLayouts(encodingType);
+              if (!encodingAttr.has_value()) {
+                return success();
+              }
+              rewriter.modifyOpInPlace(sizeOfOp, [&] {
+                sizeOfOp.setEncoding(
+                    cloneWithEncoding(encodingType, encodingAttr.value()));
+              });
+              return success();
+            })
+            .Default([](auto *op) { return failure(); });
+
+    if (failed(result)) {
+      return failure();
+    }
+  }
+  return success();
+}
+} // namespace
+
+struct SpecializeEncodingsPass
+    : public impl::SpecializeEncodingsPassBase<SpecializeEncodingsPass> {
+  void runOnOperation() override {
+    ModuleOp moduleOp = getOperation();
+    auto usedDialects = gatherUsedDialectInterfaces<
+        IREE::Stream::AffinityAnalysisDialectInterface>(moduleOp);
+    if (usedDialects.size() != 1) {
+      moduleOp.emitError("expected only one dialect implementing "
+                         "AffinityAnalysisDialectInterface");
+      return signalPassFailure();
+    }
+
+    llvm::MapVector<StringRef, IREE::Stream::ExecutableOp> executableOps;
```

Comment:
I feel like this is better:
```suggestion
llvm::MapVector<StringRef, IREE::Stream::ExecutableOp> executableOps(
    llvm::map_range(moduleOp.getOps<IREE::Stream::ExecutableOp>(),
                    [](auto op) { return std::make_pair(op.getName(), op); }));

```